### PR TITLE
Allow Side-effectful Abstract Functions in Pure Additional Functions

### PR DESCRIPTION
**Pull request by sebmarkbage**
TL;DR this tracks any reachable object or binding that a side-effectful function could mutate as having an unknown value after that call.

## Use Case

In the React compiler we see a lot of examples like this:

```js
return <div>{obj.unknown.toString()}</div>
```

Where we're calling an abstract function, or even calling a generic abstract value due to unknown being abstract.

This is an example that we currently use Flow types for. Flow lets us know that this is a number and therefore toString is known to be pure. However, fully plugging in Flow will be limiting and hard. Flow is also overkill for our use case. We still can't compile away the function call since we'll have to leave a residual call to toString anyway.

Now we could assume that all calls are pure, however, we also see a lot of patterns like this:

```js
var children = [];
obj.items.forEach(item => children.push(<Item item={item} />);
var data = knownFunction(obj.data);
return <div><Foo>{children}</Foo><Bar data={data} /></div>;
```

In this case the children array gets locally mutated. If we knew the `forEach` function here was an array (as opposed to some other `forEach` side-effectful method), then we know what it could do. Even if we knew this would deal with a loop over an abstract bound. But this assumes that we even know this is an array.

It would be sufficient for us to just leave this call in an treat the children as abstract. Even with best effort, that is more or less what's going to happen anyway.

We don't want to completely bail out on the whole function because we want to know the fact that this always returns a `<div>` with a `<Foo />` and a `<Bar />` in it so that we can continue with those. We can also inline the knownFunction call.

## Theory

Once we're inside a "pure function" such as a React component or additional function, there can be mutations.

Such a function can only mutate newly created objects and bindings, and read immutable data. That's not technically enforced but that's what the whole principle builds on.

This means that any abstract call to a function that was passed in, or lives in global scope, cannot have side-effects on anything except what was passed in as an argument (or `this`).

This PR introducing a new concept of tainting objects and bindings as abstract values after they've been passed to an impure function call.

After they have been tainted all mutations to them are tracked through a generator.

Any subsequent operations on these objects or bindings behave like access to abstract values.

## Implementation

When we enter evaluation of a "pure function" we track that scope and track all newly created objects within it as "impure".

When an abstract function call happens, I visit all arguments that are in the impure list recursively. I stop when I hit an object not on the list, since that can't be mutated.

For objects, I mark them as "tainted", reset all properties and make it partial. That way all subsequent property accesses are unknown. The abstract function call could've done anything with it, including deleting all properties.

Currently I always serialize tainted objects as empty, but I set all the properties to what they were at the first call.

Once an object is tainted, any subsequent operations on it emit to the generator.

If a function is passed in, I also taint all the _newly created_ bindings that are reachable from that closure.

Reads on a tainted binding return a new derived abstract value that replays a read of that binding. To do that I had to add a way to expand `buildNode` to be able to accept `Binding` as an argument. Not just `Value`. That way the `Referentializer` can decide what name the binding needs to serialize as.

## Errors

I currently left out error handling. Normally that doesn't matter, because if the function call throws, it will bubble up to the pure function call and act like it would've anyway.

Errors only matters if there is a `try`/`catch` anywhere between the abstract call and the pure function call. That should be solvable but even this case is so rare that it is probably fine to fatal in that case.

## Follow Up

Once we have function calls in place, we can later follow up by unlocking other operations that currently only fatals because we don't know if they will invoke side-effectful functions.

### CODE REVIEW

**Comment by hermanventer** [comment]
In the CS world the word "taint" is strongly associated with data that could be controlled by an outside agent (attacker) and thus present a security threat if it flows to unsafe sinks.

In the compiler world, the word "dirty" is more usually associated with objects that may have been modified. It would be less confusing to adopt "dirty" in the place of "taint" for this work.

File: src/realm.js [hunk 1]
```diff
@@ -368,6 +371,23 @@ export class Realm {
...
 
+  // Evaluate a context as if it won't have any side-effects outside of any objects
+  // that it created itself. This promises that any abstract functions inside of it
+  // also won't have effects on any objects or bindings that weren't created in this
+  // call.
+  evaluatePure<T>(f: () => T) {
+    let savedImpureObjects = this.impureObjects;
+    // Track all objects (including function closures) created during
+    // this call. This will be used to make the assumption that every
+    // *other* object is unchange (pure).
```
Comment (hermanventer) [inline src/realm.js]: unchange -> unchanged

File: src/realm.js [hunk 2]
```diff
@@ -368,6 +371,23 @@ export class Realm {
     this.$GlobalEnv.environmentRecord.DeleteBinding(name);
   }
 
+  // Evaluate a context as if it won't have any side-effects outside of any objects
+  // that it created itself. This promises that any abstract functions inside of it
+  // also won't have effects on any objects or bindings that weren't created in this
+  // call.
+  evaluatePure<T>(f: () => T) {
+    let savedImpureObjects = this.impureObjects;
```
Comment (hermanventer) [inline src/realm.js]: Perhaps these objects should be termed "leaked objects" since they leak into the outside world and then effectively become abstract.

Comment (sebmarkbage) [reply src/realm.js]: They are not necessarily going to get returned from the pure function so they won't always leak that way. They also won't always leak into the called abstract function.

They only have the potential to leak. Maybe "leakable". That is also a bit confusing because there other objects than these passed to the abstract function too. They don't "leak" because they were already in the outside world.

File: src/environment.js [hunk 3]
```diff
@@ -63,9 +63,17 @@ import {
   ThrowIfMightHaveBeenDeleted,
 } from "./methods/index.js";
 import * as t from "babel-types";
+import { TypesDomain, ValuesDomain } from "./domains/index.js";
 
 const sourceMap = require("source-map");
 
+export function deriveGetBinding(realm: Realm, binding: Binding) {
+  let types = new TypesDomain(Value);
```
Comment (hermanventer) [inline src/environment.js]: You may as well use TypesDomain.topVal in this case.

File: src/utils/taint.js [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+import { ResidualHeapInspector } from "../serializer/ResidualHeapInspector.js";
+
+function emitAllProperties(realm: Realm, O: ObjectValue) {
+  invariant(realm.generator, "a generator must exist while tracking a pure function");
+  let generator = realm.generator;
+  invariant(
+    O !== realm.$GlobalObject,
+    "we should not be able to get here because the global object is never mutated in a pure function."
+  );
+  // Temporary hack
```
Comment (hermanventer) [inline src/utils/taint.js]: Issue number needed here.

File: src/utils/taint.js [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+  let insp = new ResidualHeapInspector(realm, (null: any));
```
Comment (hermanventer) [inline src/utils/taint.js]: Sneaking in null where it is not expected is brittle and really not OK. Either fix ResidualHeapInspector to expect null, or pass in an actual logger here.

File: src/utils/taint.js [hunk 6]
```diff
@@ -0,0 +1,398 @@
...
+  for (let [P, propertyBinding] of O.properties) {
+    if (insp.canIgnoreProperty(O, P)) {
```
Comment (hermanventer) [inline src/utils/taint.js]: Making a new inspector object just so that you can call _canIgnoreProperty with a cache that you are going to throw away seems painful. Better would be to make _canIgnoreProperty into a public function.

File: src/utils/taint.js [hunk 7]
```diff
@@ -0,0 +1,398 @@
...
+      continue;
+    }
+    let desc = propertyBinding.descriptor;
+    if (desc === undefined) continue; // deleted
+    if (IsDataDescriptor(realm, desc) && desc.configurable && desc.enumerable && desc.writable) {
+      let descValue = desc.value || realm.intrinsics.undefined;
+      invariant(descValue instanceof Value);
+      generator.emitPropertyAssignment(O, P, descValue);
+    } else {
+      generator.emitDefineProperty(O, P, desc);
```
Comment (hermanventer) [inline src/utils/taint.js]: It this safe to do if the property is not configurable?

File: src/utils/taint.js [hunk 8]
```diff
@@ -0,0 +1,398 @@
...
+    }
+  }
+  // TODO: Emit symbol properties, prototype and other fields as well.
+}
+
+class TaintedValueVisitor {
```
Comment (hermanventer) [inline src/utils/taint.js]: A comment explaining the purpose of the class would be nice here. I find "Visitor" to be a bit a bit general and somewhat misleading. It seems that the real purpose is to call emitAllProperties on all objects that are reachable from a root. The code would be much more readable if the class name directly suggested that this is what it is doing.

File: src/utils/taint.js [hunk 9]
```diff
@@ -0,0 +1,398 @@
...
+  realm: Realm;
+  // Values to visit if they're reachable.
```
Comment (hermanventer) [inline src/utils/taint.js]: The comment says "Value", the code says "ObjectValue". Perhaps they should agreed.

File: src/utils/taint.js [hunk 10]
```diff
@@ -0,0 +1,398 @@
...
+  impureObjects: Set<ObjectValue>;
+  // Values that has been visited.
+  visitedValues: Set<Value>;
```
Comment (hermanventer) [inline src/utils/taint.js]: The code implies that only ObjectValue instances will ever be added to this set. The code may well be wrong.

File: src/utils/taint.js [hunk 11]
```diff
@@ -0,0 +1,398 @@
...
+
+  constructor(realm: Realm, impureObjects: Set<ObjectValue>) {
+    this.realm = realm;
+    this.impureObjects = impureObjects;
+    this.visitedValues = new Set();
+  }
+
+  _mark(val: Value): boolean {
```
Comment (hermanventer) [inline src/utils/taint.js]: I know this fits in with the other visitors, but I really dislike the name "_mark". It seems to me that this method should really be named "mustVisit".

File: src/utils/taint.js [hunk 12]
```diff
@@ -0,0 +1,398 @@
...
+        invariant(kind === "Object", `Object of kind ${kind} is not supported in calls to abstract functions.`);
+        invariant(
+          this.$ParameterMap === undefined,
+          `Arguments object is not supported in calls to abstract functions.`
+        );
+        return;
+    }
+  }
+
+  visitValueSymbol(val: SymbolValue): void {
```
Comment (hermanventer) [inline src/utils/taint.js]: This is never called.

File: src/utils/taint.js [hunk 13]
```diff
@@ -0,0 +1,398 @@
...
+    }
+  }
+
+  visitValue(val: Value): void {
+    if (val instanceof AbstractValue) {
+      if (this._mark(val)) this.visitAbstractValue(val);
+    } else if (val.isIntrinsic()) {
+      // All intrinsic values exist from the beginning of time...
+      // ...except for a few that come into existance as templates for abstract objects (TODO #882).
+      this._mark(val);
```
Comment (hermanventer) [inline src/utils/taint.js]: There seems to be very little point in call this._mark(val) here.

File: src/utils/taint.js [hunk 14]
```diff
@@ -0,0 +1,398 @@
...
+  }
+
+  visitObjectProperties(obj: ObjectValue, kind?: ObjectKind): void {
+    // visit properties
+    for (let [symbol, propertyBinding] of obj.symbols) {
+      invariant(propertyBinding);
+      let desc = propertyBinding.descriptor;
+      if (desc === undefined) continue; //deleted
+      this.visitDescriptor(desc);
+      this.visitValue(symbol);
```
Comment (hermanventer) [inline src/utils/taint.js]: This appears to be a no-op.

File: src/utils/taint.js [hunk 15]
```diff
@@ -0,0 +1,398 @@
...
+  visitedValues: Set<Value>;
+
+  constructor(realm: Realm, impureObjects: Set<ObjectValue>) {
+    this.realm = realm;
+    this.impureObjects = impureObjects;
+    this.visitedValues = new Set();
+  }
+
+  _mark(val: Value): boolean {
+    if (!this.impureObjects.has((val: any))) return false;
```
Comment (hermanventer) [inline src/utils/taint.js]: Avoid any like the Plague.

File: src/utils/taint.js [hunk 16]
```diff
@@ -0,0 +1,398 @@
...
+
+  visitAbstractValue(val: AbstractValue): void {
+    for (let i = 0, n = val.args.length; i < n; i++) {
+      this.visitValue(val.args[i]);
+    }
+  }
+
+  visitValue(val: Value): void {
+    if (val instanceof AbstractValue) {
+      if (this._mark(val)) this.visitAbstractValue(val);
```
Comment (hermanventer) [inline src/utils/taint.js]: _mark will never return true for an abstract value.

File: src/values/ObjectValue.js [hunk 17]
```diff
@@ -290,6 +299,19 @@ export default class ObjectValue extends ConcreteValue {
     return this._isPartial.value;
   }
 
+  makeTainted(): void {
+    this._isTainted = this.$Realm.intrinsics.true;
+    // Clear all properties. This object is now completely unknown.
+    this.properties.clear(); // TODO: Track that this happens so that it can be undone by evaluateForEffects
```
Comment (NTillmann) [inline src/values/ObjectValue.js]: This is an important TODO that should be addressed before landing.

File: src/values/ObjectValue.js [hunk 18]
```diff
@@ -290,6 +299,19 @@ export default class ObjectValue extends ConcreteValue {
...
+    // TODO: Clear symbol properties, prototype and other fields as well.
+    this.makePartial();
+    this.makeSimple(); // TODO: Don't assume it is once property access no longer fatal.
```
Comment (NTillmann) [inline src/values/ObjectValue.js]: "once property access no longer fatal". Do you have an idea how to handle this? I don't have a clear vision for this problem.

File: src/utils/taint.js [hunk 19]
```diff
@@ -0,0 +1,398 @@
...
+import { IsDataDescriptor, TestIntegrityLevel } from "../methods/index.js";
+import type { BabelNodeExpression } from "babel-types";
+import invariant from "../invariant.js";
+
+import { ResidualHeapInspector } from "../serializer/ResidualHeapInspector.js";
+
+function emitAllProperties(realm: Realm, O: ObjectValue) {
+  invariant(realm.generator, "a generator must exist while tracking a pure function");
+  let generator = realm.generator;
+  invariant(
```
Comment (NTillmann) [inline src/utils/taint.js]: Why is that invariant true? Wouldn't it get violated when I simply pass `global` as an explicit argument to a abstract function? What's so special about `$GlobalObject`, what about other built-ins (thingies with an `.intrinsicName`)?

File: src/values/ObjectValue.js [hunk 20]
```diff
@@ -230,6 +236,9 @@ export default class ObjectValue extends ConcreteValue {
   // partial objects
   _isPartial: BooleanValue;
 
+  // tainted objects
+  _isTainted: BooleanValue;
```
Comment (NTillmann) [inline src/values/ObjectValue.js]: I think there's a conceptual issue in Prepack around these tracked internal properties (https://github.com/facebook/prepack/issues/1124): The formal type should really be `BooleanValue | AbstractValue`, as merging diverged control-flow paths where a value became tainted on one side, but not on the other, would result in a conditionally tainted value. Same issue applies to `_isPartial` if applied later after construction. So, while not a new issue, we shouldn't compound the problem going forward. I think this may add a big new layer of complexity to the implementation and reasoning...

File: src/realm.js (hunk 2 above)
Comment (NTillmann) [reply src/realm.js]: +1 for "leaked objects" terminology. Functions are "pure", but I've never heard of "(im)pure" being applied objects.

"They are not necessarily going to get returned from the pure function so they won't always leak that way." Being returned is not a required, just being accessible by a function is sufficient to be considered as "leaked".

As far as I can see "leakable" is really "leaked".

File: src/realm.js [hunk 21]
```diff
@@ -205,6 +207,7 @@ export class Realm {
   modifiedBindings: void | Bindings;
   modifiedProperties: void | PropertyBindings;
   createdObjects: void | CreatedObjects;
+  impureObjects: void | ImpureObjects;
```
Comment (NTillmann) [inline src/realm.js]: So what is the meaning of `impureObjects`? Reading through the code... newly created objects within the most enclosing `evaluatePure` scope which were exposed to abstract function calls?

File: src/realm.js (hunk 2 above)
Comment (NTillmann) [reply src/realm.js]: +1 for "leaked objects" terminology. Being returns is not a requirement for "leaked".
"That is also a bit confusing because there other objects than these passed to the abstract function too." Indeed. How about "createdLeakedObjects"?

File: src/environment.js [hunk 22]
```diff
@@ -104,6 +112,8 @@ export type Binding = {
   environment: EnvironmentRecord,
   name: string,
   isGlobal: boolean,
+  isTainted: boolean,
```
Comment (NTillmann) [inline src/environment.js]: How about `hasLeaked` to stay with the leaking terminology?

File: src/utils/generator.js [hunk 23]
```diff
@@ -130,16 +134,29 @@ export class Generator {
...
       buildNode: ([]) =>
         t.expressionStatement(t.unaryExpression("delete", this.preludeGenerator.globalReference(key, !strictMode))),
     });
   }
 
+  emitBindingAssignment(binding: Binding, value: Value) {
+    this._addEntry({
+      args: [value],
+      bindings: [binding],
+      buildNode: ([valueNode, bindingNode]) =>
```
Comment (NTillmann) [inline src/utils/generator.js]: The `bindings` don't really have to be wired through `_addEntry` / `GeneratorEntry`. `context` is an extra argument passed to `buildNode`, but yet not consumed here. If you add that parameter you, you can just emit the call to `context.serializeBinding(bindingArg)` from within this `buildNode` function, instead of doing it in `_addEntry`. Also, I suggest to rename `serializeBinding` to something like `getReferentializedBindingNode`, as it doesn't actually serialize an actual value.

**Comment by NTillmann** [comment]
"Errors only matters if there is a `try`/`catch` anywhere between the abstract call and the pure function call. That should be solvable but even this case is so rare that it is probably fine to fatal in that case." That should be addressed before landing the pull request, or at least a TODO/issue be created.

File: src/realm.js (hunk 2 above)
Comment (sebmarkbage) [reply src/realm.js]: I'm hesitant to call them "leaked" since they won't all have the "hasLeaked" flag.

`objectsTrackedForLeaks`?

File: src/utils/taint.js (hunk 19 above)
Comment (sebmarkbage) [reply src/utils/taint.js]: If you pass global as an argument to an abstract function it won't be newly created so the visitor should have bailed out before you get here.

There isn't inherently special about global vs other intrinsics necessarily. However, if you do hit the global it'll pretty much reach the whole heap so it's better to bail out before it goes into some infinite loop.

We could check other intrinsics but since the intrinsic name is useful for other things (e.g. in the Node CLI there are a lot more intrinsics created than just built-in objects). It seems sketchy to assume something very specific that doesn't really have anything to do with the fact that it has an intrinsic name. E.g. I could call a function and have that object return an intrinsic object that can be recreated by calling it again.

The global object is one that is at least inherently to have been created before the additional function. There are more on that list but this is probably enough to help debugging.

File: src/utils/taint.js (hunk 7 above)
Comment (sebmarkbage) [reply src/utils/taint.js]: This is emitted on a newly created object so it won't be. Basically it serializes its initial state here, but it can have multiple initial states if there are several branches. I'm not very happy with this strategy though.

(If anything, the property assignment above isn't safe if `Object.prototype` has properties that are non-writable, but we assume they don't.)

File: src/utils/taint.js (hunk 8 above)
Comment (sebmarkbage) [reply src/utils/taint.js]: It is also flagging them as ~tainted~ leaked.

File: src/utils/taint.js (hunk 10 above)
Comment (sebmarkbage) [reply src/utils/taint.js]: AbstractValues can end up here too. Used to quickly bailout from traversing through them.

File: src/utils/taint.js (hunk 11 above)
Comment (sebmarkbage) [reply src/utils/taint.js]: `mustVisitOnce`?

File: src/values/ObjectValue.js (hunk 18 above)
Comment (sebmarkbage) [reply src/values/ObjectValue.js]: The idea I had was that invoking an unknown getter/setter is just the same as invoking an unknown function. So if you do that on a leaked object, it would generate an effect just like calling a function on it would. The object would "leak" again so any properties that are known at that time would be reset again.

For setters, it would also leak the value passed into the setter.

File: src/environment.js (hunk 3 above)
Comment (sebmarkbage) [reply src/environment.js]: Ok.

File: src/environment.js (hunk 22 above)
Comment (sebmarkbage) [reply src/environment.js]: Yea, in this case that seems pretty straightforward. Although this suffers from the same problem as the object flag. It really should vary between branches and get joined.

File: src/values/ObjectValue.js (hunk 20 above)
Comment (sebmarkbage) [reply src/values/ObjectValue.js]: Yea, I'm not really sure about what to do about this one. It is nice to avoid having to taint other branches when only one is unknown. However, if we keep this diverged path, we need to deal with it in the Binding case too.

File: src/values/ObjectValue.js (hunk 17 above)
Comment (sebmarkbage) [reply src/values/ObjectValue.js]: Yup.

File: src/utils/taint.js (hunk 16 above)
Comment (sebmarkbage) [reply src/utils/taint.js]: Yea this is surfacing a bug in `_mark` which I'll fix by only bailing out if it's an `ObjectValue` and newly created.

File: src/utils/taint.js (hunk 12 above)
Comment (sebmarkbage) [reply src/utils/taint.js]: Yea. Not sure about this. The description could be abstract, which could have objects as their arguments, but since there is no way for code to access those, I think it is safe to never visit symbols. Will remove this path.

File: src/realm.js (hunk 21 above)
Comment (sebmarkbage) [reply src/realm.js]: Newly created objects within evaluatePure. They may not have been exposed to abstract function calls yet. These are the objects that have the potential to become leaked.

File: src/utils/leak.js [hunk 24]
```diff
@@ -0,0 +1,395 @@
...
+  ObjectValue,
+  NativeFunctionValue,
+} from "../values/index.js";
+import { IsDataDescriptor, TestIntegrityLevel } from "../methods/index.js";
+import type { BabelNodeExpression } from "babel-types";
+import invariant from "../invariant.js";
+
+import { ResidualHeapInspector } from "../serializer/ResidualHeapInspector.js";
+
+function emitAllProperties(realm: Realm, O: ObjectValue) {
```
Comment (NTillmann) [inline src/utils/leak.js]: This function still seems a bit underdeveloped.

**Comment by sebmarkbage** [comment]
I think the model is in a good landable state. My next steps are:

1) Deal with the try/catch scenario mentioned in the original post. My current thinking is to just issue a recoverable error when that happens.
2) Add unit tests for the scenarios addressed by each individual commit. I can't actually run them without fixing the serializer though.
3) Fix the referentializer so that I can properly serialize bindings captured by leaked closures. This one I'll need some help with but it doesn't seem to fundamentally influence anything about the abstract evaluation and interpreter part of this.

**Comment by sebmarkbage** [comment]
One big question here is whether we want to enable this for all abstract functions in additional functions, since it can possibly regress if those are considered pure today.

I'd propose that we introduce a new way to create an abstract _pure_ function. E.g. `__abstract('purefunction')` or something. Then those will still be considered pure. Normal `__abstract('function')` or calls to unknown abstract types would be considered side-effectful and can only be called in the context of an additional function or React component, in which case the arguments would leak.

@cblappert Do we rely on a lot of calls to abstract pure functions in RN models?

**Comment by NTillmann** [comment]
also see here regarding pure vs. non-pure functions: #705

File: src/utils/leak.js [hunk 25]
```diff
@@ -0,0 +1,445 @@
...
+  "AssignmentExpression|UpdateExpression"(path: BabelTraversePath, state: LeakedFunctionInfo) {
+    let doesRead = path.node.operator !== "=";
+    for (let name in path.getBindingIdentifiers()) {
+      visitName(path, state, name, doesRead, true);
+    }
+  },
+};
+
+function getLeakedFunctionInfo(value: FunctionValue) {
+  // TODO: This should really be cached on a per AST basis in case we have
```
Comment (NTillmann) [inline src/utils/leak.js]: Indeed. Babel visitors are crazy expensive.

**Comment by NTillmann** [comment]
LGTM. Before formally signing off...
- Do you want to remove the [RFC] tag?
- Can you fix the test issue?
- @cblappert, what do you think is the nature of the abstract functions in our models? Any regression risks?

**Comment by sebmarkbage** [comment]
Ok. I'd love to land this PR to avoid merge conflicts but I don't feel confident turning it on in all abstract functions yet, until #1245 is fixed. So maybe I just revert the part that opts into a "pure tree" which triggers the leak mechanism?

**Comment by sebmarkbage** [comment]
I'll just add a flag so that we can turn it on/off. That way I can at least add the unit tests that can pass.

**Comment by sebmarkbage** [comment]
I'll try to add test cases for the ones that can be serialized. I'm blocked on #1245 for many of them since a lot of scenarios end up being serialized incorrectly due to bindings.
